go/registry/api/api.go

```diff
@@ -402,37 +403,37 @@ func VerifyRegisterNodeArgs( // nolint: gocyclo
 		}
 	}
 
-	var expectedSigner signature.PublicKey
-	if inEntityNodeList {
-		expectedSigner = n.ID
-	} else if entity.AllowEntitySignedNodes {
-		expectedSigner = entity.ID
-	} else {
-		logger.Error("RegisterNode: node registration has no valid signer",
-			"node", n,
-			"entity", entity,
-		)
-		return nil, nil, fmt.Errorf("%w: registration has no valid signer", ErrInvalidArgument)
-	}
-
-	// Validate that the node is signed by the correct signer.
-	if sigNode.Signed.Signature.SanityCheck(expectedSigner) != nil {
-		logger.Error("RegisterNode: not signed by expected signer",
+	// Descriptors will always be signed by the node identity key.
+	var expectedSigners []signature.PublicKey
+	if !sigNode.MultiSigned.IsSignedBy(n.ID) {
+		logger.Error("RegisterNode: registration not signed by node identity",
 			"signed_node", sigNode,
 			"node", n,
-			"entity", entity,
 		)
-		return nil, nil, fmt.Errorf("%w: registration not signed by expected signer", ErrInvalidArgument)
+		return nil, nil, fmt.Errorf("%w: registration not signed by node identity", ErrInvalidArgument)
+	}
+	expectedSigners = append(expectedSigners, n.ID)
+	if !inEntityNodeList && entity.AllowEntitySignedNodes {
+		// If we are using entity signing, descriptors will also be signed
+		// by the entity signing key.
+		if !sigNode.MultiSigned.IsSignedBy(entity.ID) {
+			logger.Error("RegisterNode: registration not signed by entity",
+				"signed_node", sigNode,
+				"node", n,
+			)
+			return nil, nil, fmt.Errorf("%w: registration not signed by entity", ErrInvalidArgument)
+		}
+		expectedSigners = append(expectedSigners, entity.ID)
 	}
 
-	// Checking the expiration only makes sense if this routine is
-	// validating the genesis document, since node descriptors being
-	// expired is presumably ok at runtime(?).
-	if isGenesis && n.Expiration <= uint64(epoch) {
-		logger.Error("RegisterNode: expired node in genesis",
+	// Registrations that are expired are allowed in the genesis document
+	// as expired node registrations are preserved through the debonding
+	// period of an entity.
+	if !isGenesis && n.Expiration <= uint64(epoch) {
+		logger.Error("RegisterNode: registration is expired",
 			"node", n,
 		)
-		return nil, nil, fmt.Errorf("%w: expired node in genesis", ErrInvalidArgument)
+		return nil, nil, fmt.Errorf("%w: registration is expired", ErrInvalidArgument)
```

Comment:
i know this was wrong before, but it should probably be `ErrNodeExpired`

Reply:
Eh, nothing else there does fine grained error reporting.